subsys/net/lib/coap/coap_client.c

```diff
@@ -0,0 +1,639 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(net_coap, CONFIG_COAP_LOG_LEVEL);
+
+#include <zephyr/net/socket.h>
+
+#include <zephyr/net/coap.h>
+#include <zephyr/net/coap_client.h>
+
+#define COAP_VERSION 1
+#define COAP_PATH_ELEM_DELIM '/'
+#define COAP_PATH_ELEM_QUERY '?'
+#define MAX_COAP_MSG_LEN CONFIG_COAP_ASYNC_CLIENT_BUFFER_SIZE
+#define COAP_POOL_SLEEP 500
+#define COAP_SEPARATE_TIMEOUT 6000
+#define DEFAULT_RETRY_AMOUNT 5
+
+static K_SEM_DEFINE(coap_client_recv_sem, 0, 1);
+static atomic_t coap_client_recv_active;
+static struct pollfd fds;
+static struct coap_client_request *coap_request;
+static int retry_count;
+static struct coap_packet request;
+static uint8_t request_token[COAP_TOKEN_MAX_LEN];
+static int request_tkl;
+static uint8_t send_buf[MAX_COAP_MSG_LEN];
+static uint8_t recv_buf[MAX_COAP_MSG_LEN];
+static struct coap_block_context recv_blk_ctx;
+static struct coap_block_context send_blk_ctx;
+static struct coap_pending pending;
+static struct sockaddr address;
+static socklen_t socklen;
```

Comment:
Making everything `static` like this only allows for a single client, can't we introduce a `struct coap_client` type that would allow for re-usability?

Reply:
I totally agree here, we've already been through this with other helper libs (like TFTP) - we'll eventually end up reworking the API so support multiple parallel clients.

IMO it'd be better to do it right from start, i. e. introduce a client context structure and provide it as an argument to the API functions.

Reply:
Good point, needs a bit of refactoring. In that case I think we could also drop the KConfig flag for `CONFIG_COAP_ASYNC_CLIENT`

Reply:
I made the changes now, the client uses `struct coap_client` which needs to be initialized with `coap_client_init` before using. This allows using multiple clients and nothing should be statically allocated in the client code now.